core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -576,19 +581,22 @@ object Doc {
   /**
    * Represents an optimistic rendering (on the left) as well as a
    * fallback rendering (on the right) if the first line of the left
-   * is too long.
-   *
-   * There is an additional invariant on Union: `flatten(a) == flatten(b)`.
-   *
-   * By construction all `Union` nodes have this property; to preserve
+   * is too long.  More precisely, in `Union(a, b)`, we have the
+   * invariants:
+   *
+   * - `a` consists entirely of {`Empty`, `Text`, `Concat`} nodes.
```

Comment:
why do we need this? This is very strong. We aren't allowing deferred nodes and it winds up creating costs in fill which you made the PR for in: #115 

Can we document why we want to require this? Ideally, we keep the requirements as weak as possible.

I don't know why the requirement that the left side has length >= the right isn't enough. I'd sooner want to remove invariants (such as the invariant the the `a.flatten == b.flatten`) than add them, but I guess the latter was added so we could efficiently implement flatten at all (though I suppose we could define `Union(a, b).flatten == Union(a.flatten, b.flatten)` if we didn't require `a.flatten == b.flatten`.

Reply:
Yeah, I just misinterpreted the comment in the code.  I was trying to understand the code and be careful what invariants I needed to maintain before making the big `FlatAlt` change.  Clearly I overshot the mark here.  I'll make the requirement you intended more clear here.